File: 42-trapping-rain-water/trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        if not height:
            return 0
        
        l, r = 0, len(height) - 1
        lmax, rmax = height[l], height[r]
        res = 0

        while l < r:
            lmax, rmax = max(height[l], lmax), max(height[r], rmax)

            if lmax <= rmax:
                res += lmax - height[l]
                l += 1
            else:
                res += rmax - height[r]
                r -= 1
        
        return res

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        if (height.length == 0) {
            return 0;
        }
        
        int l = 0, r = height.length - 1;
        int lmax = height[l], rmax = height[r];
        int res = 0;

        while (l < r) {
            lmax = Math.max(height[l], lmax);
            rmax = Math.max(height[r], rmax);

            if (lmax <= rmax) {
                res += lmax - height[l];
                l++;
            } else {
                res += rmax - height[r];
                r--;
            }
        }
        
        return res;
    }
}

